### PR TITLE
Added == and != operators to ValueTuple.cs. Closes #10416

### LINKED ISSUE #10416: Add == and != operators to ValueTuple.cs
After the following discussion on https://github.com/dotnet/coreclr/issues/6520#issuecomment-236152874 we discovered ValueTuple did not implement the == and != operators.

I think this is an oversight, it looks pretty natural to me to write:
```csharp
(name, age) = GetPersonTuple();
if ((name, age) == ("Jhon", 13))
{
    // ...
}
```
OR
```csharp
tuple = GetPersonTuple();
if (tuple == ("Jhon", 13))
{
    // ...
}
```

Personally, I think that if the performance cost is not too high I might even compare variables using the tuple notation. It is easier to understand and to read than the `&&` syntax:
```csharp
var age = AskAge();
var name = AskName();

if((age, name) == (17, "Jhon"))
{
    // ...
}
```
instead of:
```csharp
var age = AskAge();
var name = AskName();

if(age == 17 && name == "Jhon")
{
    // ...
}
```

This might overlap with the Tuple Patterns but I think those were removed from C# 7. Having to call .Equals seems very ugly to me.
```csharp
tuple = GetPersonTuple();
if (tuple is ("Jhon", 13))
{
    // ...
}
```

### PR DESCRIPTION
**Pull request by OzieGamma**
I know I am not supposed to send a PR to master with public API changes but I could not find ValueTuple.cs and felt like this would be a good opportunity to help the .NET OSS effort !

Related issue: https://github.com/dotnet/corefx/issues/10416

### CODE REVIEW

**Comment by jcouv** [comment]
Thanks @OzieGamma for the PR!

@stephentoub @terrajobst The change looks good to me. I'll let you comment on public API question.
In particular, this introduces a difference between `ValueTuple` and `Tuple`. Should we add similar operators to `System.Tuple` too?

FYI for @KevinRansom @gafter @VSadov

**Comment by OzieGamma** [comment]
I don't think we (you ?) can add == and != to `System.Tuple`. Since it is a reference type, == is already defined. That would introduce breaking changes in the public API.

This compiles fine:
```csharp
Tuple<string> a = new Tuple<string>("");
Tuple<string> b = new Tuple<string>("");

if (a == b)
{
}
```

On the other hand, as mentioned above, users will potentially use ValueTuple a LOT. Not having == or != might be frustrating ...

*NB: I haven't played around with the new language features yet. I don't know in practice how likely I will need/want `==` or `!=`*

**Comment by jcouv** [comment]
@OzieGamma You're right about backward compatibility issue on `System.Tuple`.
Let me check with some folks and get back to you on this PR.

**Comment by gafter** [comment]
I have two main concerns with adding operator== and operator!= to System.ValueTuple. Both of them boil down to the assertion that these operators should act as if operating using == on the underlying elements. But you can’t program that into the ValueTuple type, because the implementation of the generic type doesn’t know what operator== to apply to the elements. For example

1.	Using object.Equals(x, y) isn’t correct, because it isn’t the same as operator==. To take an example from our platform, double.NaN.Equals(double.Nan), but !(double.NaN==double.NaN). So (double.NaN == double.NaN) would give a different result from ((double.NaN, 0) == (double.NaN, 0)) if we implement the latter using object.Equals(); the former would be false while the latter would be true. I believe that is not what we want.
2.	We would want the == operator on tuples to apply whenever an underlying operator== can be found to apply on the elements, pointwise. If we do it in the language+compiler (now or in the future), we can make ((short)1, (byte)2) == ((long)1, (int)2) be true (not a compile-time error because no operator== can be found). However, there is no way to express that in source.

In short, I believe that adding support for operator== and operator!= to the sources of the ValueTuple libraries would be locking us in to the wrong semantics for what we would want the behavior to be.

**Comment by OzieGamma** [comment]
@gafter Your points make complete sense. I'll propose this on the Roslyn repo then.

**Comment by OzieGamma** [comment]
I
